File: iOS/samples/iOSSample/Auth Provider/MSATokenCache.m
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

#import "MSATokenCache.h"
#import "MSATokenRequest.h"

static NSString* const MsaOfflineAccessScope = @"wl.offline_access";

static NSString* const JsonTokenKey = @"refresh_token";
static NSString* const JsonExpirationKey = @"expires";

// Max number of times to try to refresh a token through transient failures
static const NSUInteger MsaTokenRefreshMaxRetries = 3;

// How quickly to retry refresh token refreshes on transient failure; 30 minutes.
static const int64_t MsaRefreshTokenRetryInterval = 30 * 60;

// How quickly to retry access token refreshes on transient failure; 3 minutes.
static const int64_t MsaAccessTokenRetryInterval = 3 * 60;

// How long a refresh token is expected to last without expiring; 10 days.
static const NSTimeInterval MsaRefreshTokenExpirationInterval = 10 * 24 * 60 * 60;

// Time from expiration at which a refresh token is considered 'close to expiring'; 7 days.
// (This value is intended to be aggressive and keep the refresh token relatively far from expiry)
static const NSTimeInterval MsaRefreshTokenCloseToExpiryInterval = 7 * 24 * 60 * 60;

// Time from expiration at which an access token is considered 'close to expiring'; 5 minutes.
static const NSTimeInterval MsaAccessTokenCloseToExpiryInterval = 5 * 60;

// @brief Private helper class, encapsulates a single MSA token to be cached, and how to refresh it.
@interface MSATokenCacheItem : NSObject
+ (nullable instancetype)cacheItemWithToken:(nonnull NSString*)token
                                  expiresIn:(NSTimeInterval)expiry
                                refreshWith:(nonnull MSATokenRequest*)refreshRequest
                                     parent:(nonnull MSATokenCache*)parent;
- (nullable instancetype)initWithToken:(nonnull NSString*)token
                             expiresIn:(NSTimeInterval)expiry
                           refreshWith:(nonnull MSATokenRequest*)refreshRequest
                                parent:(nonnull MSATokenCache*)parent;

// Asynchronously fetches the token held by this item, refreshing it if necessary.
- (void)getTokenAsync:(nonnull void (^)(NSString* _Nullable token))callback;

@property(readwrite, nonnull, nonatomic, copy) NSString* token;
@property(readwrite, nonnull, nonatomic, strong) NSDate* expirationDate;
@property(readwrite, nonnull, nonatomic, strong) MSATokenRequest* refreshRequest;
@property(readwrite, nonnull, nonatomic, strong) MSATokenCache* parent;
@property(readonly, nonatomic) NSTimeInterval closeToExpiryInterval;
@property(readonly, nonatomic) int64_t retryInterval;

// Private helper for refreshing this token. Only to be used by this class and its subclass.
// Returns the refresh token needed to refresh the token held by this item.
// For access tokens, this gets the refresh token held by the cache.
// For refresh tokens, just return the currently-held token.
- (void)getRefreshTokenAsync:(nonnull void (^)(NSString* _Nullable token))callback;

// Private helper for refreshing this token. Only to be used by this class and its subclass.
// For access tokens, sets the new token and expiration.
// For refresh tokens, marks current access tokens as expired, and caches the refresh token in persistent storage.
- (void)onSuccessfulRefresh:(nonnull MSATokenRequestResult*)result;

@end

// @brief Subclass of MSATokenCacheItem for refresh tokens
@interface MSARefreshTokenCacheItem : MSATokenCacheItem
+ (nullable instancetype)loadSavedRefreshTokenWithParent:(nonnull MSATokenCache*)parent;
- (void)saveRefreshToken;
@end

// MSATokenCache privates
@interface MSATokenCache ()
@property(readonly, nonnull, nonatomic, copy) NSString* clientId;
@property(readonly, nonnull, nonatomic, strong) NSMutableDictionary<NSString*, MSATokenCacheItem*>* cachedAccessTokens; // keyed on scopes
@property(readwrite, nullable, nonatomic, strong) MSARefreshTokenCacheItem* cachedRefreshToken;
- (void)markAccessTokensExpired;
@end

@implementation MSATokenCacheItem

+ (instancetype)cacheItemWithToken:(NSString*)token
                         expiresIn:(NSTimeInterval)expiry
                       refreshWith:(MSATokenRequest*)refreshRequest
                            parent:(MSATokenCache*)parent
{
    return [[self alloc] initWithToken:token expiresIn:expiry refreshWith:refreshRequest parent:parent];
}

- (instancetype)initWithToken:(NSString*)token
                    expiresIn:(NSTimeInterval)expiry
                  refreshWith:(MSATokenRequest*)refreshRequest
                       parent:(MSATokenCache*)parent
{
    if (self = [super init])
    {
        _token = [token copy];
        _expirationDate = [NSDate dateWithTimeIntervalSinceNow:expiry];
        _refreshRequest = refreshRequest;
        _parent = parent;
    }
    return self;
}

- (void)getTokenAsync:(void (^)(NSString*))callback
{
    [self getTokenAsync:callback maxRetries:MsaTokenRefreshMaxRetries];
}

- (void)getTokenAsync:(void (^)(NSString*))callback maxRetries:(NSUInteger)maxRetries
{
    if ([_expirationDate timeIntervalSinceNow] >= self.closeToExpiryInterval)
    {
        // If expiration date is sufficiently far away
        callback(self.token);
    }
    else
    {
        // If expired or close to it, get the refresh token and attempt to refresh with it
        [self getRefreshTokenAsync:^void(NSString* refreshToken) {
            if (!refreshToken)
            {
                // Unable to get the refresh token even after retrying
                // Consider as a permanent failure and call back with no tokens
                NSLog(@"Unable to get refresh token. Cancelling refresh and removing all tokens from cache.");
                [_parent clearTokens];
                callback(nil);
            }

            NSLog(@"Refreshing token...");
            [_refreshRequest
                requestAsyncWithToken:refreshToken
                             callback:^void(MSATokenRequestResult* result) {
                                 switch (result.status)
                                 {
                                 case MSATokenRequestStatusSuccess:
                                 {
                                     [self onSuccessfulRefresh:result];
                                     callback(self.token);
                                     break;
                                 }
                                 case MSATokenRequestStatusTransientFailure:
                                 {
                                     if (maxRetries > 0)
                                     {
                                         // Retry the refresh
                                         NSLog(@"Encountered transient error when refreshing token, retrying in %lld seconds...",
                                             self.retryInterval);
                                         dispatch_time_t retryTime = dispatch_time(DISPATCH_TIME_NOW, self.retryInterval * NSEC_PER_SEC);
                                         dispatch_after(retryTime, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0),
                                             ^{ [self getTokenAsync:callback maxRetries:(maxRetries - 1)]; });
                                     }
                                     else
                                     {
                                         // Reached max number of retries
                                         NSLog(@"Reached max number of retries for refreshing token.");
                                         callback(nil);
                                     }
                                     break;
                                 }
                                 default: // PermanentFailure
                                 {
                                     NSLog(@"Permanent error occurred while refreshing token. Clearing the cache...");
                                     [_parent clearTokens];
                                     [_parent.delegate onTokenCachePermanentFailure];
                                     callback(nil);
                                     break;
                                 }
                                 }
                             }];
        }];
    }
}

- (NSTimeInterval)closeToExpiryInterval
{
    return MsaAccessTokenCloseToExpiryInterval; // Base class expects access tokens
}

- (int64_t)retryInterval
{
    return MsaAccessTokenRetryInterval; // Base class expects access tokens
}

- (void)getRefreshTokenAsync:(void (^)(NSString*))callback
{
    [_parent getRefreshTokenAsync:callback]; // Base class expects access tokens, grab the refresh token from the parent
}

- (void)onSuccessfulRefresh:(MSATokenRequestResult*)result
{
    @synchronized(self)
    {
        NSString* newToken = result.accessToken;
        NSAssert(newToken.length > 0, @"UNEXPECTED: Refresh access token succeeded but access token was empty.");

        NSLog(@"Successfully refreshed access token.");
        self.token = newToken;
        self.expirationDate = [NSDate dateWithTimeIntervalSinceNow:result.expiresIn];
    }
}

@end

@implementation MSARefreshTokenCacheItem

+ (instancetype)loadSavedRefreshTokenWithParent:(MSATokenCache*)parent
{
    NSLog(@"Loading refresh token from keychain...");

    // clang-format off
    NSDictionary* keychainMatchQuery = @{
        (id) kSecClass          : (id) kSecClassGenericPassword,
        (id) kSecAttrGeneric    : parent.clientId,
        (id) kSecMatchLimit     : (id) kSecMatchLimitOne, // Only match one keychain item
        (id) kSecReturnData     : @YES                    // Return the data itself rather than a ref
    };
    // clang-format on

    CFTypeRef keychainItems = NULL;
    OSStatus keychainStatus = SecItemCopyMatching((CFDictionaryRef)keychainMatchQuery, &keychainItems);
    if (keychainStatus == errSecItemNotFound)
    {
        NSLog(@"No refresh token found in keychain.");
        return nil;
    }
    else if (keychainStatus != errSecSuccess)
    {
        NSLog(@"Unable to load refresh token from keychain with OSStatus %d", (int)keychainStatus);
        return nil;
    }

    NSError* jsonError = nil;
    CFDataRef tokenData = (CFDataRef)keychainItems;
    id deserializedTokenData = [NSJSONSerialization JSONObjectWithData:(__bridge NSData*)tokenData options:0 error:&jsonError];
    if (jsonError)
    {
        NSLog(@"Encountered JSON error \'%@\' while trying to load refresh token from keychain.", jsonError);
        return nil;
    }
    else if (![deserializedTokenData isKindOfClass:[NSDictionary class]])
    {
        NSLog(@"Loaded refresh token data from keychain was in an unexpected format. Will not load.");
        return nil;
    }

    NSDictionary* tokenDict = (NSDictionary*)deserializedTokenData;
    NSString* loadedRefreshToken = (NSString*)(tokenDict[JsonTokenKey]);

    NSDateFormatter* dateFormatter = [NSDateFormatter new];
    dateFormatter.dateStyle = NSDateFormatterFullStyle;
    dateFormatter.timeStyle = NSDateFormatterFullStyle;
    NSDate* loadedRefreshTokenExpiry = [dateFormatter dateFromString:(NSString*)(tokenDict[JsonExpirationKey])];
    if (!loadedRefreshToken || !loadedRefreshTokenExpiry)
    {
        NSLog(@"Loaded refresh token data from keychain was incomplete or corrupted.");
        return nil;
    }

    NSTimeInterval timeUntilExpiration = [loadedRefreshTokenExpiry timeIntervalSinceDate:[NSDate date]];
    MSATokenRequest* refreshRequest = [MSATokenRequest tokenRequestWithClientId:parent.clientId
                                                                      grantType:MsaTokenRequestGrantTypeRefresh
                                                                          scope:MsaOfflineAccessScope
                                                                    redirectUri:nil];
    MSARefreshTokenCacheItem* ret =
        [self cacheItemWithToken:loadedRefreshToken expiresIn:timeUntilExpiration refreshWith:refreshRequest parent:parent];

    NSLog(@"Successfully loaded refresh token from keychain.");
    return ret;
}

- (void)saveRefreshToken
{
    NSLog(@"Saving refresh token to keychain...");

    NSDateFormatter* dateFormatter = [NSDateFormatter new];
    dateFormatter.dateStyle = NSDateFormatterFullStyle;
    dateFormatter.timeStyle = NSDateFormatterFullStyle;
    NSDictionary* tokenDict = @{ JsonTokenKey : self.token, JsonExpirationKey : [dateFormatter stringFromDate:self.expirationDate] };

    NSError* jsonError = nil;
    NSData* tokenData = [NSJSONSerialization dataWithJSONObject:tokenDict options:0 error:&jsonError];
    if (jsonError)
    {
        NSLog(@"Encountered JSON error \'%@\' while trying to save refresh token to keychain. Will not save.", jsonError);
        return;
    }

    // clang-format off
    NSDictionary* keychainSearchQuery = @{
        (id) kSecClass          : (id) kSecClassGenericPassword,
        (id) kSecAttrGeneric    : self.parent.clientId
    };
    // clang-format on

    OSStatus keychainStatus = SecItemUpdate((CFDictionaryRef)keychainSearchQuery, (CFDictionaryRef) @{ (id) kSecValueData : tokenData });
    if (keychainStatus == errSecItemNotFound)
    {
        // After a device restart, this keychain item is only accessible after the device is unlocked at least once.
        // This keychain item is not migrated when restoring a backup from another device.
        id accessAttribute = (id)kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly;

        NSMutableDictionary* keychainAddQuery = [keychainSearchQuery mutableCopy];
        [keychainAddQuery addEntriesFromDictionary:@{ (id) kSecAttrAccessible : accessAttribute, (id) kSecValueData : tokenData }];
        keychainStatus = SecItemAdd((CFDictionaryRef)keychainAddQuery, NULL);
    }

    if (keychainStatus != errSecSuccess)
    {
        NSLog(@"Failed to save refresh token data to keychain with OSStatus %d.", (int)keychainStatus);
    }

    NSLog(@"Successfully saved refresh token data to keychain.");
}

- (NSTimeInterval)closeToExpiryInterval
{
    return MsaRefreshTokenCloseToExpiryInterval;
}

- (int64_t)retryInterval
{
    return MsaRefreshTokenRetryInterval;
}

- (void)getRefreshTokenAsync:(void (^)(NSString*))callback
{
    callback(self.token); // Since this cache item holds a refresh token, just return it
}

- (void)onSuccessfulRefresh:(MSATokenRequestResult*)result
{
    @synchronized(self)
    {
        NSString* newToken = result.refreshToken;
        NSAssert(newToken.length > 0, @"UNEXPECTED: Refresh refresh token succeeded but access token was empty.");

        NSLog(@"Successfully refreshed refresh token.");
        self.token = newToken;
        self.expirationDate = [NSDate dateWithTimeIntervalSinceNow:MsaRefreshTokenExpirationInterval];
        [self saveRefreshToken];
        [self.parent markAccessTokensExpired];
    }
}

@end

// MSATokenCache implementation
@implementation MSATokenCache

+ (instancetype)cacheWithClientId:(NSString*)clientId delegate:(id<MSATokenCacheDelegate>)delegate
{
    return [[self alloc] initWithClientId:clientId delegate:delegate];
}

- (instancetype)initWithClientId:(NSString*)clientId delegate:(id<MSATokenCacheDelegate>)delegate
{
    if (self = [super init])
    {
        _clientId = [clientId copy];
        _delegate = delegate;
        _cachedAccessTokens = [NSMutableDictionary<NSString*, MSATokenCacheItem*> new];
    }
    return self;
}

- (void)setRefreshToken:(NSString*)refreshToken
{
    MSATokenRequest* refreshRequest = [MSATokenRequest tokenRequestWithClientId:_clientId
                                                                      grantType:MsaTokenRequestGrantTypeRefresh
                                                                          scope:MsaOfflineAccessScope
                                                                    redirectUri:nil];
    @synchronized(self)
    {
        _cachedRefreshToken = [MSARefreshTokenCacheItem cacheItemWithToken:refreshToken
                                                                 expiresIn:MsaRefreshTokenExpirationInterval
                                                               refreshWith:refreshRequest
                                                                    parent:self];
        [_cachedRefreshToken saveRefreshToken];
        [self markAccessTokensExpired];
    }
}

- (void)setAccessToken:(NSString*)accessToken forScope:(NSString*)scope expiresIn:(NSTimeInterval)expiry
{
    MSATokenRequest* refreshRequest =
        [MSATokenRequest tokenRequestWithClientId:_clientId grantType:MsaTokenRequestGrantTypeRefresh scope:scope redirectUri:nil];

    @synchronized(self)
    {
        [_cachedAccessTokens
            setValue:[MSATokenCacheItem cacheItemWithToken:accessToken expiresIn:expiry refreshWith:refreshRequest parent:self]
              forKey:scope];
    }
}

- (void)getRefreshTokenAsync:(void (^)(NSString*))callback
{
    @synchronized(self)
    {
        if (_cachedRefreshToken)
        {
            [_cachedRefreshToken getTokenAsync:callback];
        }
        else
        {
            callback(nil);
        }
    }
}

- (void)getAccessTokenForScopeAsync:(NSString*)scope callback:(void (^)(NSString*))callback
{
    @synchronized(self)
    {
        MSATokenCacheItem* item = [_cachedAccessTokens valueForKey:scope];
        if (item)
        {
            [item getTokenAsync:callback];
        }
        else
        {
            callback(nil);
        }
    }
}

- (NSArray<NSString*>*)allScopes
{
    return [_cachedAccessTokens allKeys];
}

- (BOOL)loadSavedRefreshToken
{
    MSARefreshTokenCacheItem* loadedRefreshToken = [MSARefreshTokenCacheItem loadSavedRefreshTokenWithParent:self];
    if (loadedRefreshToken)
    {
        if ([loadedRefreshToken.expirationDate compare:[NSDate date]] != NSOrderedDescending)
        {
            NSLog(@"Refresh token loaded from keychain was expired. Ignoring.");
            return NO;
        }

        @synchronized(self)
        {
            _cachedRefreshToken = loadedRefreshToken;
            [self markAllTokensExpired]; // Force a refresh on everything on first use
        }
    }

    return (loadedRefreshToken != nil);
}

- (void)clearTokens
{
    NSLog(@"Clearing token data from cache...");

    @synchronized(self)
    {
        [_cachedAccessTokens removeAllObjects];
        _cachedRefreshToken = nil;
    }

    // clang-format off
    NSDictionary* keychainDeleteQuery = @{
        (id) kSecClass          : (id) kSecClassGenericPassword,
        (id) kSecAttrGeneric    : _clientId
    };
    // clang-format on

    OSStatus keychainStatus = SecItemDelete((CFDictionaryRef)keychainDeleteQuery);
    if (keychainStatus != errSecSuccess)
    {
        NSLog(@"Unable to clear token data from keychain with OSStatus %d. Data might still be loaded on next run.", (int)keychainStatus);
    }

    NSLog(@"Done clearing token data from cache.");
}

- (void)markAccessTokensExpired
{
    @synchronized(self)
    {
        for (MSATokenCacheItem* cachedAccessToken in _cachedAccessTokens.allValues)
        {
            cachedAccessToken.expirationDate = [NSDate distantPast];
        }
    }
}

- (void)markAllTokensExpired
{
    @synchronized(self)
    {
        _cachedRefreshToken.expirationDate = [NSDate distantPast];
        [self markAccessTokensExpired];
    }
}

@end

File: Android/samples/account-provider-sample/src/main/java/com/microsoft/connecteddevices/sampleaccountproviders/MSATokenCache.java
//
// Copyright (c) Microsoft Corporation. All rights reserved.
//

package com.microsoft.connecteddevices.signinhelpers;

import android.content.Context;
import android.support.annotation.Keep;
import android.util.ArrayMap;
import android.util.Log;
import android.content.SharedPreferences;

import com.microsoft.connecteddevices.AsyncOperation;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.text.DateFormat;
import java.text.ParseException;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collection;
import java.util.Date;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * Caches MSA access and refresh tokens, automatically refreshing them as needed when fetching from the cache.
 * Cached refresh tokens are persisted across sessions.
 */
@Keep
final class MSATokenCache {
    private static final String TAG = MSATokenCache.class.getName();

    // Max number of times to try to refresh a token through transient failures
    private static final int TOKEN_REFRESH_MAX_RETRIES = 3;

    // How quickly to retry refreshing a token when encountering a transient failure
    private static final long MSA_REFRESH_TOKEN_RETRY_SECONDS = 30 * 60; // 30 minutes
    private static final long MSA_ACCESS_TOKEN_RETRY_SECONDS = 3 * 60;   // 3 minutes

    // How long it takes a refresh token to expire
    private static final int MSA_REFRESH_TOKEN_EXPIRATION_SECONDS = 10 * 24 * 60 * 60; // 10 days

    // How long before expiry to consider a token in need of a refresh.
    // (MSA_REFRESH_TOKEN_CLOSE_TO_EXPIRY_SECONDS is intended to be aggressive and keep the refresh token relatively far from expiry)
    private static final int MSA_REFRESH_TOKEN_CLOSE_TO_EXPIRY_SECONDS = 7 * 24 * 60 * 60; // 7 days
    private static final int MSA_ACCESS_TOKEN_CLOSE_TO_EXPIRY_SECONDS = 5 * 60;            // 5 minutes

    private static final String MSA_OFFLINE_ACCESS_SCOPE = "wl.offline_access";
    private static final String GUID_ID_KEY = "GUID_ID_KEY";

    private static final ScheduledExecutorService sRetryExecutor = Executors.newSingleThreadScheduledExecutor();

    /**
     * Helper function. Returns a Date n seconds from now.
     */
    private static final Date getDateSecondsAfterNow(int seconds) {
        return getDateSecondsAfter(null, seconds);
    }

    /**
     * Helper function. Returns a Date n seconds after date.
     */
    private static final Date getDateSecondsAfter(Date date, int seconds) {
        Calendar calendar = Calendar.getInstance(); // sets time to current
        if (date != null) {
            calendar.setTime(date);
        }
        calendar.add(Calendar.SECOND, seconds);
        return calendar.getTime();
    }

    /**
     * Private helper class wrapping a cached access token. Responsible for refreshing it on-demand.
     */
    private class MSATokenCacheItem {
        protected String mToken;
        protected Date mCloseToExpirationDate; // Actual expiration date is used less than this, so just cache this instead
        protected final MSATokenRequest mRefreshRequest;

        public MSATokenCacheItem(String token, int expiresInSeconds, MSATokenRequest refreshRequest) {
            mToken = token;
            mCloseToExpirationDate = getDateSecondsAfterNow(expiresInSeconds - getCloseToExpirySeconds());
            mRefreshRequest = refreshRequest;
        }

        /**
         * Returns the number of seconds before expiry that this token is considered in need of a refresh.
         */
        protected int getCloseToExpirySeconds() {
            return MSA_ACCESS_TOKEN_CLOSE_TO_EXPIRY_SECONDS; // Base class expects access tokens
        }

        /**
         * Returns the number of seconds to wait before retrying, when a refresh fails with a transient error.
         */
        protected long getRetrySeconds() {
            return MSA_ACCESS_TOKEN_RETRY_SECONDS; // Base class expects access tokens
        }

        /**
         * Returns the refresh token to use to refresh the token held by this item.
         * For access tokens, this gets the refresh token held by the cache.
         * For refresh tokens, this just returns the currently-held token.
         */
        protected AsyncOperation<String> getRefreshTokenAsync() {
            return MSATokenCache.this.getRefreshTokenAsync();
        }

        /**
         * Steps to complete after a successful refresh.
         * For access tokens, sets the new token and new expiration.
         * For refresh tokens, marks current access tokens as expired, and caches the refresh token in persistent storage.
         */
        protected synchronized void onSuccessfulRefresh(MSATokenRequest.Result result) {
            Log.i(TAG, "Successfully refreshed access token.");
            mToken = result.getAccessToken();
            mCloseToExpirationDate = getDateSecondsAfterNow(result.getExpiresIn() - getCloseToExpirySeconds());
        }

        /**
         * Private helper - asynchronously fetches the token held by this item.
         * If the token is close to expiry, refreshes it first.
         * If this refresh fails due to transient error, recursively retries up to remainingRetries times to refresh.
         *
         * @param operation         AsyncOperation to return the token on
         * @param remainingRetries  number of times to retry refreshing, in the case of transient error
         * @return the operation that was passed in
         */
        private AsyncOperation<String> _getTokenAsyncInternal(final AsyncOperation<String> operation, final int remainingRetries) {
            if (!needsRefresh()) {
                operation.complete(mToken); // Already have a non-stale token, can just return with it
                return operation;
            }

            getRefreshTokenAsync()
                    .thenComposeAsync((AsyncOperation.ResultFunction<String, AsyncOperation<MSATokenRequest.Result>>) refreshToken -> mRefreshRequest.requestAsync(refreshToken))
                    .thenAcceptAsync((AsyncOperation.ResultConsumer<MSATokenRequest.Result>) result -> {
                        switch (result.getStatus()) {
                            case SUCCESS:
                                onSuccessfulRefresh(result);
                                operation.complete(mToken);
                                break;

                            case TRANSIENT_FAILURE:
                                // Recursively retry the refresh, if there are still remaining retries
                                if (remainingRetries <= 0) {
                                    Log.e(TAG, "Reached max number of retries for refreshing token.");
                                    operation.complete(null);

                                } else {
                                    Log.i(TAG, "Transient error while refreshing token, retrying in " + getRetrySeconds() + "seconds...");
                                    sRetryExecutor.schedule(() -> {
                                        _getTokenAsyncInternal(operation, remainingRetries - 1);
                                    }, getRetrySeconds(), TimeUnit.SECONDS);
                                }
                                break;

                            default: // PERMANENT_FAILURE
                                Log.e(TAG, "Permanent error occurred while refreshing token.");
                                MSATokenCache.this.onPermanentFailure();
                                operation.complete(null);
                                break;
                        }
                    });

            return operation;
        }

        /**
         * Asynchronously fetches the token held by this item, refreshing it if necessary.
         */
        public AsyncOperation<String> getTokenAsync() {
            AsyncOperation<String> ret = new AsyncOperation<String>();
            return _getTokenAsyncInternal(ret, TOKEN_REFRESH_MAX_RETRIES);
        }

        public boolean needsRefresh() {
            return mCloseToExpirationDate.before(new Date());
        }

        public boolean isExpired() {
            return getDateSecondsAfter(mCloseToExpirationDate, getCloseToExpirySeconds()).before(new Date());
        }

        public synchronized void markExpired() {
            mCloseToExpirationDate = new Date(0); // Start of epoch
        }
    }

    /**
     * Private helper class wrapping a cached refresh token. Responsible for refreshing it on demand. Can translate to/from json format.
     */
    private final class MSARefreshTokenCacheItem extends MSATokenCacheItem {
        private static final String JSON_TOKEN_KEY = "refresh_token";
        private static final String JSON_EXPIRATION_KEY = "expires";

        public MSARefreshTokenCacheItem(String token, int expiresInSeconds, MSATokenRequest refreshRequest) {
            super(token, expiresInSeconds, refreshRequest);
        }

        public MSARefreshTokenCacheItem(JSONObject json) throws IOException, JSONException, ParseException {
            super(null, 0, new MSATokenRequest(mClientId, MSATokenRequest.GrantType.REFRESH, MSA_OFFLINE_ACCESS_SCOPE, null));

            mToken = json.optString(JSON_TOKEN_KEY);
            String dateString = json.optString(JSON_EXPIRATION_KEY);
            if (mToken == null || dateString == null) {
                throw new IOException("Saved refresh token was improperly formatted.");
            }

            Date expirationDate = DateFormat.getDateTimeInstance().parse(dateString);
            mCloseToExpirationDate = getDateSecondsAfter(expirationDate, -MSA_REFRESH_TOKEN_CLOSE_TO_EXPIRY_SECONDS);
        }

        protected int getCloseToExpirySeconds() {
            return MSA_REFRESH_TOKEN_CLOSE_TO_EXPIRY_SECONDS;
        }

        protected long getRetrySeconds() {
            return MSA_REFRESH_TOKEN_RETRY_SECONDS;
        }

        protected AsyncOperation<String> getRefreshTokenAsync() {
            return AsyncOperation.completedFuture(mToken);
        }

        protected synchronized void onSuccessfulRefresh(MSATokenRequest.Result result) {
            Log.i(TAG, "Successfully refreshed refresh token.");
            mToken = result.getRefreshToken();
            mCloseToExpirationDate =
                    getDateSecondsAfterNow(MSA_REFRESH_TOKEN_EXPIRATION_SECONDS - MSA_REFRESH_TOKEN_CLOSE_TO_EXPIRY_SECONDS);
            MSATokenCache.this.markAccessTokensExpired();
            MSATokenCache.this.trySaveRefreshToken();
        }

        public synchronized JSONObject toJSON() throws JSONException {
            // Get the actual expiration date
            Date expirationDate = getDateSecondsAfter(mCloseToExpirationDate, MSA_REFRESH_TOKEN_CLOSE_TO_EXPIRY_SECONDS);

            JSONObject ret = new JSONObject();
            ret.put(JSON_TOKEN_KEY, mToken);
            ret.put(JSON_EXPIRATION_KEY, DateFormat.getDateTimeInstance().format(expirationDate));
            return ret;
        }
    }

    /**
     * Provides callbacks when the cache encounters a permanent failure and has to wipe its state.
     */
    public static interface Listener { void onTokenCachePermanentFailure(); }

    private final String mClientId;
    private final Context mContext;

    private MSARefreshTokenCacheItem mCachedRefreshToken = null;
    private final Map<String, MSATokenCacheItem> mCachedAccessTokens = new ArrayMap<>();

    private final Collection<Listener> mListeners = new ArrayList<>();

    public MSATokenCache(String clientId, Context context) {
        mClientId = clientId;
        mContext = context;
    }

    /**
     * Returns a file in application-specific storage that's used to persist the refresh token across sessions.
     */
    private File getRefreshTokenSaveFile() throws IOException {
        Context appContext = mContext.getApplicationContext();
        File appDirectory = appContext.getDir(appContext.getPackageName(), Context.MODE_PRIVATE);
        if (appDirectory == null) {
            throw new IOException("Could not access app directory.");
        }

        return new File(appDirectory, "samplemsaaccountprovider.dat");
    }

    /**
     * Tries to save the current refresh token to persistent storage.
     */
    private void trySaveRefreshToken() {
        Log.i(TAG, "Trying to save refresh token...");
        try {
            File file = getRefreshTokenSaveFile();
            JSONObject json = file.exists() ? new JSONObject(IOUtil.readUTF8Stream(new FileInputStream(file))) : new JSONObject();

            json.put(mClientId, mCachedRefreshToken.toJSON());
            IOUtil.writeUTF8Stream(new FileOutputStream(file), json.toString());

            Log.i(TAG, "Saved refresh token.");

        } catch (IOException | JSONException e) {
            Log.e(TAG, "Exception while saving refresh token. \"" + e.getLocalizedMessage() + "\" Will not save.");
        }
    }

    /**
     * Tries to read a saved refresh token from persistent storage, and return it as an MSARefreshTokenItem.
     */
    private MSARefreshTokenCacheItem tryReadSavedRefreshToken() {
        Log.i(TAG, "Trying to read saved refresh token...");
        try {
            File file = getRefreshTokenSaveFile();

            if (!file.exists()) {
                Log.i(TAG, "No saved refresh token was found.");
                return null;
            }

            JSONObject json = new JSONObject(IOUtil.readUTF8Stream(new FileInputStream(file)));
            JSONObject innerJson = json.optJSONObject(mClientId);

            if (innerJson == null) {
                Log.i(TAG, "Could not read saved refresh token.");
                return null;
            }

            Log.i(TAG, "Read saved refresh token.");
            return new MSARefreshTokenCacheItem(innerJson);

        } catch (IOException | JSONException | ParseException e) {
            Log.e(TAG, "Exception reading saved refresh token. \"" + e.getLocalizedMessage() + "\"");
            return null;
        }
    }

    /**
     * Tries to delete the saved refresh token for this app in persistent storage.
     */
    private void tryClearSavedRefreshToken() {
        Log.i(TAG, "Trying to delete saved refresh token...");
        try {
            File file = getRefreshTokenSaveFile();
            if (!file.exists()) {
                Log.i(TAG, "No saved refresh token was found.");
                return;
            }

            try {
                // Try to remove just a section of the json corresponding to client id
                JSONObject json = new JSONObject(IOUtil.readUTF8Stream(new FileInputStream(file)));
                json.remove(mClientId);

                if (json.length() <= 0) {
                    file.delete(); // Just delete the file if the json would be empty
                } else {
                    IOUtil.writeUTF8Stream(new FileOutputStream(file), json.toString());
                }
            } catch (JSONException e) {
                // Failed to parse the json, just delete everything
                file.delete();
            }

            Log.i(TAG, "Deleted saved refresh token.");

        } catch (IOException e) { Log.e(TAG, "Failed to delete saved refresh token. \"" + e.getLocalizedMessage() + "\""); }
    }

    /**
     * Marks access tokens as expired, such that a refresh is performed before returning, when the access token is next requested.
     */
    private synchronized void markAccessTokensExpired() {
        for (MSATokenCacheItem cachedAccessToken : mCachedAccessTokens.values()) {
            cachedAccessToken.markExpired();
        }
    }

    /**
     * Calls back any listeners that the cache has encountered a permanent failure, and that they should perform any needed error-handling.
     */
    private void onPermanentFailure() {
        clearTokens();
        for (Listener listener : mListeners) {
            listener.onTokenCachePermanentFailure();
        }
    }

    public void setRefreshToken(String refreshToken) {
        MSATokenRequest refreshRequest = new MSATokenRequest(mClientId, MSATokenRequest.GrantType.REFRESH, MSA_OFFLINE_ACCESS_SCOPE, null);

        synchronized (this) {
            mCachedRefreshToken = new MSARefreshTokenCacheItem(refreshToken, MSA_REFRESH_TOKEN_EXPIRATION_SECONDS, refreshRequest);
            markAccessTokensExpired();
            trySaveRefreshToken();
        }
    }

    public void setAccessToken(String accessToken, String scope, int expiresInSeconds) {
        MSATokenRequest refreshRequest = new MSATokenRequest(mClientId, MSATokenRequest.GrantType.REFRESH, scope, null);

        synchronized (this) {
            mCachedAccessTokens.put(scope, new MSATokenCacheItem(accessToken, expiresInSeconds, refreshRequest));
        }
    }

    public synchronized AsyncOperation<String> getRefreshTokenAsync() {
        if (mCachedRefreshToken != null) {
            return mCachedRefreshToken.getTokenAsync();
        } else {
            return AsyncOperation.completedFuture(null);
        }
    }

    public synchronized AsyncOperation<String> getAccessTokenAsync(String scope) {
        MSATokenCacheItem cachedAccessToken = mCachedAccessTokens.get(scope);
        if (cachedAccessToken != null) {
            return cachedAccessToken.getTokenAsync();
        } else {
            return AsyncOperation.completedFuture(null);
        }
    }

    public void saveAccountId(String id) {
        // Get the shared preferences
        SharedPreferences preferences = mContext.getSharedPreferences(mContext.getPackageName(), Context.MODE_PRIVATE);

        // Save the given ID to the shared preferences
        preferences.edit().putString(GUID_ID_KEY, id).apply();
    }

    public String readSavedAccountId() {
        // Get the shared preferences
        SharedPreferences preferences = mContext.getSharedPreferences(mContext.getPackageName(), Context.MODE_PRIVATE);

        // Grab the value of the key with a default value of empty string
        String id = preferences.getString(GUID_ID_KEY, "");
        // Check that we found a value and not the default value
        if (id.isEmpty()) {
            Log.e(TAG, "readSavedAccountId failed to get the ID");
        }
        return id;
    }

    public synchronized void addListener(Listener listener) {
        mListeners.add(listener);
    }

    public synchronized void removeListener(Listener listener) {
        mListeners.remove(listener);
    }

    public Set<String> allScopes() {
        return mCachedAccessTokens.keySet();
    }

    /**
     * Tries to load a saved refresh token from disk. If successful, the loaded refresh token is used as this cache's refresh token.
     * @return Whether a saved refresh token was loaded successfully.
     */
    public boolean loadSavedRefreshToken() {
        Log.i(TAG, "Trying to load saved refresh token...");
        MSARefreshTokenCacheItem savedRefreshToken = tryReadSavedRefreshToken();

        if (savedRefreshToken == null) {
            Log.i(TAG, "Failed to load saved refresh token.");
            return false;
        }

        if (savedRefreshToken.isExpired()) {
            Log.i(TAG, "Read saved refresh token, but was expired. Ignoring.");
            return false;
        }

        Log.i(TAG, "Successfully loaded saved refresh token.");
        mCachedRefreshToken = savedRefreshToken;
        markAllTokensExpired(); // Force a refresh on everything on first use
        return true;
    }

    /**
     * Clears all tokens from the cache, and any saved refresh tokens belonging to this app in persistent storage.
     */
    public synchronized void clearTokens() {
        mCachedAccessTokens.clear();
        mCachedRefreshToken = null;
        tryClearSavedRefreshToken();
    }

    /**
     * Marks all tokens as expired, such that a refresh is performed before returning, when a token is next requested.
     */
    public synchronized void markAllTokensExpired() {
        mCachedRefreshToken.markExpired();
        markAccessTokensExpired();
    }
}
